dipy/denoise/fast_lpca.pyx

```diff
@@ -0,0 +1,286 @@
+    n = x.shape[0]
+    if utype == 0:
+        for i in range(n):
+            for j in range(n):
+                out[i, j] = x[i] * x[j]
+    elif utype == 1:
+        for i in range(n):
+            for j in range(n):
+                out[i, j] += x[i] * x[j]
+    elif utype == -1:
+        for i in range(n):
+            for j in range(n):
+                out[i, j] -= x[i] * x[j]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_vector(double[:] out, double[:] x, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, n
+    n = x.shape[0]
+    if utype == 0:
+        for i in range(n):
+            out[i] = x[i]
+    elif utype == 1:
+        for i in range(n):
+            out[i] += x[i]
+    elif utype == -1:
+        for i in range(n):
+            out[i] -= x[i]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_matrix(double[:, :] out, double[:, :] cov, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, j, n
+    n = cov.shape[0]
+    if utype == 0:
+        for i in range(n):
+            for j in range(n):
+                out[i, j] = cov[i, j]
+    elif utype == 1:
+        for i in range(n):
+            for j in range(n):
+                out[i, j] += cov[i, j]
+    elif utype == -1:
+        for i in range(n):
+            for j in range(n):
+                out[i, j] -= cov[i, j]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def fast_lpca(double[:, :, :, :] I, int radius, double[:, :, :] sigma):
+    '''
+    Local PCA Based Denoising of Diffusion Datasets
+
+    Parameters
+    ----------
+    I : A 4D array which is to be denoised
+
+    sigma : 4D array
+        standard deviation of the noise estimated from the data
+        If the sigma is a float or a 3D array one muct convert it
+        into a 4D array before passing
+
+    radius : The radius of the local patch to
+             be taken around each voxel
+    Returns
+    -------
+
+    denoised_arr : 4D array
+        this is the denoised array of the same size as that of
+        arr (input data)
+
+    References
+    ----------
+    [1] Diffusion Weighted Image Denoising Using Overcomplete Local PCA
+        Manjon JV, Coupe P, Concha L, Buades A, Collins DL
+
+    [2] On the computation of integrals over fixed-size rectangles of
+        arbitrary dimension
+        Omar Ocegueda, Oscar Dalmau, Eleftherios Garyfallidis,
+        Maxime Descoteaux, Mariano Rivera
+
+    '''
+
+    cdef:
+        cnp.npy_intp n0 = I.shape[0]
+        cnp.npy_intp n1 = I.shape[1]
+        cnp.npy_intp n2 = I.shape[2]
+        cnp.npy_intp ndiff = I.shape[3]
+        cnp.npy_intp m = 1 + (2 * radius)
+        cnp.npy_intp nsamples = m * m * m
+        cnp.npy_intp cur_i, prev_i, i, j, k, ii
+        cnp.npy_intp i0, j0, k0, p, q, r, s
+        double l0norm
+        double[:, :, :, :, :] T = np.zeros((2, n1, n2, ndiff, ndiff))
```

Comment:
It may be tempting to refactor everything to plain arrays. I just tried but allocating memory resulted in an extreme overhead, so unless we reserve the temporary memory as one single memory block it won't improve performance.